File: ios/PersonalSiteTests/MotionBehaviorTests.swift
import Testing

@testable import PersonalSite

struct MotionBehaviorTests {
    @Test func appShellUsesFivePrimaryTabsAndTwoFollowingModes() {
        #expect(AppTab.allCases == [.home, .aiNews, .following, .works, .ask])
        #expect(FollowingMode.allCases == [.recommendations, .openSource])
    }

    @Test func careerTimelineShowsTheFourPublicMilestones() {
        #expect(CareerTimelineView.years == ["2014", "2019", "2023", "2026"])
        #expect(CareerTimelineView.arrowCount == 3)
        #expect(ProfileHeaderContent.expandedTimelineTrailingInset == 18)
        #expect(ProfileHeaderContent.expandedLinksTrailingInset == 10)
        #expect(
            ProfileHeaderContent.expandedTimelineTrailingInset
                > ProfileHeaderContent.expandedLinksTrailingInset
        )
        #expect(CareerTimelineView.fontSize == 12.2)
        #expect(CareerTimelineView.opacity(for: 3) > CareerTimelineView.opacity(for: 0))
        #expect(CareerTimelineView.runnerWidth == 18)
        #expect(CareerTimelineView.stageHeight >= CareerTimelineView.runnerHeight)
        #expect(CareerTimelineView.runnerCenterX(progress: 1, width: 200) > 200)
        #expect(!CareerTimelineView.shouldRenderRunner(
            shouldAnimateEntrance: false,
            reduceMotion: false
        ))
    }

    @Test func bioFinalSnapshotIsCompleteChineseCopy() {
        let snapshot = BioAnimationSnapshot.final
        #expect(snapshot.titleVisibleCount == BioView.chineseTitle.count)
        #expect(snapshot.visibleCounts == BioView.profileCopy.map(\.count))
    }

    @Test func signalFieldOnlyAnimatesWhenAllActivityConditionsPass() {
        #expect(SignalFieldActivity.shouldAnimate(reduceMotion: false, sceneIsActive: true, isVisible: true))
        #expect(!SignalFieldActivity.shouldAnimate(reduceMotion: true, sceneIsActive: true, isVisible: true))
        #expect(!SignalFieldActivity.shouldAnimate(reduceMotion: false, sceneIsActive: false, isVisible: true))
        #expect(!SignalFieldActivity.shouldAnimate(reduceMotion: false, sceneIsActive: true, isVisible: false))
    }
}

File: android/app/src/test/java/com/personalsite/MotionBehaviorTests.kt
package com.personalsite

import com.personalsite.features.home.BioAnimationSnapshot
import com.personalsite.features.home.BioCopy
import com.personalsite.features.home.CareerTimelineSpec
import com.personalsite.features.home.ProfileHeaderInsets
import com.personalsite.features.home.SignalFieldActivity
import com.personalsite.features.support.AppTab
import com.personalsite.features.support.FollowingMode
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

class MotionBehaviorTests {
    @Test
    fun `appShellUsesFivePrimaryTabsAndTwoFollowingModes`() {
        assertEquals(
            listOf(AppTab.HOME, AppTab.AI_NEWS, AppTab.FOLLOWING, AppTab.WORKS, AppTab.ASK),
            AppTab.entries,
        )
        assertEquals(
            listOf(FollowingMode.RECOMMENDATIONS, FollowingMode.OPEN_SOURCE),
            FollowingMode.entries,
        )
    }

    @Test
    fun `careerTimelineShowsTheFourPublicMilestones`() {
        assertEquals(listOf("2014", "2019", "2023", "2026"), CareerTimelineSpec.years)
        assertEquals(3, CareerTimelineSpec.arrowCount)
        assertEquals(18f, ProfileHeaderInsets.EXPANDED_TIMELINE_TRAILING)
        assertEquals(10f, ProfileHeaderInsets.EXPANDED_LINKS_TRAILING)
        assertTrue(
            ProfileHeaderInsets.EXPANDED_TIMELINE_TRAILING > ProfileHeaderInsets.EXPANDED_LINKS_TRAILING
        )
        assertEquals(12.2, CareerTimelineSpec.FONT_SIZE, 1e-9)
        assertTrue(CareerTimelineSpec.opacity(3) > CareerTimelineSpec.opacity(0))
        assertEquals(18f, CareerTimelineSpec.RUNNER_WIDTH)
        assertTrue(CareerTimelineSpec.STAGE_HEIGHT >= CareerTimelineSpec.RUNNER_HEIGHT)
        assertTrue(CareerTimelineSpec.runnerCenterX(progress = 1f, width = 200f) > 200f)
        assertFalse(
            CareerTimelineSpec.shouldRenderRunner(
                shouldAnimateEntrance = false,
                reduceMotion = false,
            )
        )
    }

    @Test
    fun `bioFinalSnapshotIsCompleteChineseCopy`() {
        val snapshot = BioAnimationSnapshot.final
        assertEquals(BioCopy.chineseTitle.length, snapshot.titleVisibleCount)
        assertEquals(BioCopy.profileCopy.map { it.length }, snapshot.visibleCounts)
    }

    @Test
    fun `signalFieldOnlyAnimatesWhenAllActivityConditionsPass`() {
        assertTrue(SignalFieldActivity.shouldAnimate(reduceMotion = false, sceneIsActive = true, isVisible = true))
        assertFalse(SignalFieldActivity.shouldAnimate(reduceMotion = true, sceneIsActive = true, isVisible = true))
        assertFalse(SignalFieldActivity.shouldAnimate(reduceMotion = false, sceneIsActive = false, isVisible = true))
        assertFalse(SignalFieldActivity.shouldAnimate(reduceMotion = false, sceneIsActive = true, isVisible = false))
    }
}
